File: SBR obs and domains/SBR code - Python/Parser.py
import TreeNode
import xml.etree.ElementTree as ET

class Parser(object):

    def __init__(self):
        self._id_counter = 0
        self._hmap = {}
        self._goals = []
        self._recipes = []
        self._doc = None


    def parse(self, path):
        root = TreeNode.TreeNode(self.generate_ID(), "root")
        root.set_root(True)
        tree = ET.parse(path)
        self._doc = tree.getroot()
        self.read_non_terminal_letters(root)
        self.read_recipes(root)
        return root

    def read_non_terminal_letters(self, root):
        non_terminal_letters = self._doc.find("Letters").find("Non-Terminals").findall("Letter")
        for letter in non_terminal_letters:
            if (letter.get("goal") == "yes"):
                self._goals.append(letter.get("id"))
                p = TreeNode.TreeNode(self.generate_ID(), letter.get("id"))
                root.add_child(p)
                self._hmap[letter.get("index")] = p

    def read_recipes(self, root):
        recipes_as_node_list = self._doc.find("Recipes").findall("Recipe")
        for i in recipes_as_node_list:
            self._recipes.append(i)
        while self._recipes != []:
            recipe = self._recipes.pop(0)
            if root.find_by_label(recipe.get("lhs")) != None:
                p = root.find_by_label(recipe.get("lhs"))
            else:
                self._recipes.insert(len(self._recipes), recipe)
                continue
            self.read_single_recipe(recipe, p)
            self.read_order_cons(recipe)

    def read_single_recipe(self,recipe, p):
        letters = recipe.findall("Letter")
        for letter in letters:
            ID = self.generate_ID()
            child = TreeNode.TreeNode(ID,letter.get("id"))
            p.add_child(child)
            self._hmap[letter.get("index")] = child

    def read_order_cons(self, recipe):
        order = recipe.find("Order")
        if order != None:
            order_conses = order.findall("OrderCons")
            for order_cons in order_conses:
                p = self._hmap[order_cons.get("firstIndex")]
                self._hmap[order_cons.get("secondIndex")].add_seq_of(p.get_ID(), p)

    def generate_ID(self):
        self._id_counter += 1
        return self._id_counter

File: standardization/Algorithms/SBR-JAVA/src/main/java/Parser.java
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import java.io.File;
import java.io.IOException;
import java.util.*;

public class Parser {

    private int idCounter = 0;

    private Document doc;

    private HashMap<String, TreeNode> hmap = new HashMap<String, TreeNode>();

    private ArrayList<String> goals = new ArrayList<String>();

    private LinkedList<org.w3c.dom.Node> Recipes = new LinkedList<org.w3c.dom.Node>();

    private ArrayList<String> arrayOfUnauthorizedLetters = new ArrayList<>(Arrays.asList("MU","RB","S","SA","SAB","SwA","MR","ChC","CRT","DS","DT","MS"));

    public TreeNode parse(String path) throws ParserConfigurationException, SAXException, IOException {
        TreeNode root = new TreeNode("root", generateID());
        root.setRoot(true);
        File fXmlFile = new File(path);
        DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
        DocumentBuilder dBuilder = dbFactory.newDocumentBuilder();
        doc = dBuilder.parse(fXmlFile);
        doc.getDocumentElement().normalize();
        readNonTerminalLetters(root);
        readRecipes(root);
        System.out.println(root);
        return root;
}

    private void readNonTerminalLetters(TreeNode root) {
        NodeList nonTerminalLetters = ((Element) doc.getElementsByTagName("Non-Terminals").item(0)).getElementsByTagName("Letter");
        for (int i = 0; i < nonTerminalLetters.getLength(); i++) {
            if (((Element) nonTerminalLetters.item(i)).getAttribute("goal").equals("yes")) {
                goals.add(((Element) nonTerminalLetters.item(i)).getAttribute("id"));
                TreeNode p = new TreeNode(((Element) nonTerminalLetters.item(i)).getAttribute("id"), generateID());
                root.addChild(p);
                hmap.put(((Element) nonTerminalLetters.item(i)).getAttribute("index"), p);
            }
        }
    }

    private void readRecipes(TreeNode root) {
        NodeList RecipesAsNodeList = doc.getElementsByTagName("Recipe");
        for (int i = 0; i < RecipesAsNodeList.getLength(); i++) {
            Recipes.add(RecipesAsNodeList.item(i));
        }
        while (!Recipes.isEmpty()) {
            //Recipes.forEach(element -> System.out.println(((Element) element).getAttribute("lhs")));
            //System.out.println("-----------------------------");
            Element Recipe = (Element) Recipes.removeFirst();
            TreeNode p;
            if (root.findByLabel(Recipe.getAttribute("lhs")) != null) {
                p = root.findByLabel(Recipe.getAttribute("lhs"));
            } else {
                if (!arrayOfUnauthorizedLetters.contains(Recipe.getAttribute("lhs"))) {
                    Recipes.addLast(Recipe);
                }
                continue;
            }

            readSingleRecipe(Recipe, p);

            readOrderCons(Recipe);
        }
    }

    private void readSingleRecipe(Element Recipe, TreeNode p)
    {
        //parse the Letters - children's of recipe.
        int j = 0;
        Element Letter = (Element) Recipe.getElementsByTagName("Letter").item(j);
        while (Letter != null) {
            //add plan children's
            int ID = generateID();
            TreeNode child = new TreeNode((Letter.getAttribute("id")), ID);

            p.addChild(child);

            hmap.put(Letter.getAttribute("index"), child);

            //parse next letter
            j += 1;

            Letter = (Element) Recipe.getElementsByTagName("Letter").item(j);
        }
    }

    private void readOrderCons(Element Recipe) {
        int k = 0;
        Element OrderCons = (Element) Recipe.getElementsByTagName("OrderCons").item(k);

        //parse the OrderCons - seq edges.

        while (OrderCons != null) {
            //add seq edges

            TreeNode p = hmap.get(OrderCons.getAttribute("firstIndex"));
            hmap.get(OrderCons.getAttribute("secondIndex")).addSeqOf(p.getID(), p);

            //System.out.println( "second index " +  hmap.get(OrderCons.getAttribute("secondIndex")) +  "first index " + hmap.get(OrderCons.getAttribute("firstIndex")));

            k += 1;
            OrderCons = (Element) Recipe.getElementsByTagName("OrderCons").item(k);
        }
    }

    private int generateID() {
        idCounter++;
        return idCounter;
    }
}